bindings/matrix-sdk-ffi/src/api.udl

```diff
@@ -112,57 +106,179 @@ interface Room {
     [Throws=ClientError]
     string? member_display_name(string user_id);
 
-    BackwardsStream? start_live_event_listener();
-    void stop_live_event_listener();
+    void add_timeline_listener(TimelineListener listener);
+
+    // Loads older messages into the timeline.
+    //
+    // Raises an exception if there are no timeline listeners.
+    [Throws=ClientError]
+    PaginationOutcome paginate_backwards();
+
+    // Removes the timeline.
+    //
+    // Timeline items cached in memory as well as timeline listeners are dropped.
+    void remove_timeline();
 
     [Throws=ClientError]
     void send(MessageEventContent msg, string? txn_id);
 };
 
-interface BackwardsStream {
-    sequence<AnyMessage> paginate_backwards(u64 count);
+callback interface TimelineListener {
+    void on_update(TimelineDiff update);
 };
 
-interface MessageEventContent {};
+// Should really be an enum
+interface TimelineDiff {
+    TimelineChange change();
 
-interface AnyMessage {
-    TextMessage? text_message();
-    ImageMessage? image_message();
-    NoticeMessage? notice_message();
-    EmoteMessage? emote_message();
+    [Self=ByArc]
+    sequence<TimelineItem>? replace();
+    [Self=ByArc]
+    InsertAtData? insert_at();
+    [Self=ByArc]
+    UpdateAtData? update_at();
+    u32? remove_at();
+    MoveData? move();
+    [Self=ByArc]
+    TimelineItem? push();
 };
 
-interface BaseMessage {
-    string id();
-    string body();
+enum TimelineChange {
+    "Replace",
+    "InsertAt",
+    "UpdateAt",
+    "RemoveAt",
+    "Move",
+    "Push",
+    "Pop",
+    "Clear",
+};
+
+dictionary InsertAtData {
+    u32 index;
+    TimelineItem item;
+};
+
+dictionary UpdateAtData {
+    u32 index;
+    TimelineItem item;
+};
+
+dictionary MoveData {
+    u32 old_index;
+    u32 new_index;
+};
+
+interface TimelineItem {
+    [Self=ByArc]
+    EventTimelineItem? as_event();
+    [Self=ByArc]
+    VirtualTimelineItem? as_virtual();
+};
+
+interface EventTimelineItem {
+    TimelineKey key();
     string sender();
```

Comment:
Would it be possible to have this `sender` provide the displayname and avatar URL so we don't need to fetch those separately? E.g. something along the lines of
```
struct SenderDetails {
    let id: String
    let displayname: String
    let avatarURL: URL
}
```

Reply:
This was discussed a while back and decided against, partially to keep the API simpler and partially because clients would usually want to have one mapping of user ID to displayname / avatar per room that's used both for the timeline and member list.

Reply:
> one mapping of user ID to displayname / avatar per room

A counter-point to this. Currently in Element, we show the name/avatar of the user at the point in time when they sent the event. So having a single mapping wouldn't allow that to be the case.

Reply:
Yeah, that's a known trade-off, it's noted in https://hdoc.ergaster.org/IFvYg5aSSkaBizYFmVlgIA. Unfortunately I don't have a permalink to the previous discussion about this, but maybe we can just re-discuss.